File: src/main/java/org/llschall/ardwloop/motor/ProgramContainer.java
package org.llschall.ardwloop.motor;

import org.jetbrains.annotations.Nullable;
import org.llschall.ardwloop.IArdwProgram;
import org.llschall.ardwloop.structure.data.ProgramCfg;
import org.llschall.ardwloop.structure.data.SerialData;
import org.llschall.ardwloop.structure.model.Model;

import java.util.ArrayList;
import java.util.List;


public class ProgramContainer {

    public final Model model;
    private final Config config;
    private final IArdwProgram program;
    private final List<AbstractLoop> loops = new ArrayList<>();

    public ProgramContainer(IArdwProgram program) {
        this.program = program;

        Model model = new Model(this);
        model.serialMdl.program.set(new ProgramCfg(program.getId(), program.getRc(), program.getSc()));

        config = new Config(9600, model);
        this.model = model;
    }

    public void addLoop(AbstractLoop loop) {
        loops.add(loop);
    }

    public void start() {
        Clock clock = new Clock(config, loops, model);
        clock.start();
    }

    public final SerialData setupPrg(SerialData s) {
        return program.setup(s);
    }

    public final SerialData loopPrg(SerialData r) {
        return program.loop(r);
    }

    public final void postPrg(@Nullable SerialData p) {
        program.post(p);
    }

}

File: src/main/kotlin/org/llschall/ardwloop/motor/ProgramContainer.kt
package org.llschall.ardwloop.motor

import org.llschall.ardwloop.IArdwProgram
import org.llschall.ardwloop.serial.IArdwPortSelector
import org.llschall.ardwloop.serial.port.ISerialProvider
import org.llschall.ardwloop.structure.data.ProgramCfg
import org.llschall.ardwloop.structure.model.ArdwloopModel
import org.llschall.ardwloop.structure.utils.Timer
import org.llschall.ardwloop.value.SerialData

class ProgramContainer(private val program: IArdwProgram) {

    @JvmField
    val model: ArdwloopModel
    private val loops: MutableList<AbstractLoop> = ArrayList()

    init {
        val model = ArdwloopModel(this)
        model.serialMdl.program.set(
            ProgramCfg(
                program.programId,
                program.readDelayMs, program.postDelayMs,
            )
        )

        this.model = model
    }

    fun addLoop(loop: AbstractLoop) {
        loops.add(loop)
    }

    fun start(provider: ISerialProvider, baud: Int, resetPin: Int, timer: Timer, selector: IArdwPortSelector) {
        val clock = Clock(provider, timer, loops, model, selector)
        model.serialMdl.baud.set(baud)
        model.serialMdl.resetPin.set(resetPin)
        clock.start()
    }

    fun setupPrg(s: SerialData): SerialData {
        return program.ardwSetup(s)
    }

    fun loopPrg(r: SerialData?): SerialData {
        return program.ardwLoop(r)
    }

    fun postPrg(p: SerialData?) {
        program.ardwPost(p)
    }
}
